rafx-framework/src/render_features/jobs/extract/extract_job.rs

```diff
@@ -0,0 +1,210 @@
+            .render_feature_view_packet(view_index)
+    }
+
+    fn extract_render_object_instance_per_view(
+        &self,
+        view_packet: &dyn RenderFeatureViewPacket,
+        range: Range<usize>,
+    ) {
+        if range.is_empty() || !self.inner.can_extract_render_object_instance_per_view() {
+            return;
+        }
+
+        profiling::scope!(ExtractScopeNames::scope(
+            &self.scope_names.extract_render_object_instance_per_view
+        ));
+
+        let frame_packet = self.frame_packet.as_ref().unwrap();
+        let view_packet: &ViewPacket<ExtractJobEntryPointsT::FramePacketDataT> =
+            view_packet.as_concrete();
+
+        for id in range {
+            let context =
+                ExtractRenderObjectInstancePerViewContext::new(frame_packet, view_packet, id);
+            self.inner.extract_render_object_instance_per_view(&context);
+        }
+    }
+
+    fn end_per_view_extract(
+        &self,
+        view_packet: &dyn RenderFeatureViewPacket,
+    ) {
+        profiling::scope!(ExtractScopeNames::scope(
+            &self.scope_names.end_per_view_extract
+        ));
+
+        let view_packet: &ViewPacket<ExtractJobEntryPointsT::FramePacketDataT> =
+            view_packet.as_concrete();
+
+        let context = ExtractPerViewContext::new(self.frame_packet.as_ref().unwrap(), view_packet);
+        if let Some(render_objects) = self.inner.render_objects() {
+            self.inner.end_per_view_extract(&render_objects, &context);
+        } else {
+            self.inner
+                .end_per_view_extract(&self.empty_render_objects, &context);
+        }
+    }
+
+    fn end_per_frame_extract(&self) {
+        profiling::scope!(ExtractScopeNames::scope(
+            &self.scope_names.end_per_frame_extract
+        ));
+
+        let context = ExtractPerFrameContext::new(self.frame_packet.as_ref().unwrap());
+        if let Some(render_objects) = self.inner.render_objects() {
+            self.inner.end_per_frame_extract(&render_objects, &context);
+        } else {
+            self.inner
+                .end_per_frame_extract(&self.empty_render_objects, &context);
+        }
+    }
+
+    fn num_views(&self) -> usize {
+        self.view_packets().len()
+    }
+
+    fn num_render_object_instances(&self) -> usize {
+        self.render_object_instances().len()
+    }
+
+    fn take_frame_packet(&mut self) -> Box<dyn RenderFeatureFramePacket> {
+        std::mem::take(&mut self.frame_packet).unwrap()
+    }
+
+    fn feature_debug_name(&self) -> &'static str {
+        self.inner.feature_debug_name()
+    }
+
+    fn feature_index(&self) -> RenderFeatureIndex {
+        self.inner.feature_index()
+    }
+}
+
+struct ExtractScopeNames {
+    begin_per_frame_extract: String,
+    extract_render_object_instance: String,
+    extract_render_object_instance_per_view: String,
+    end_per_view_extract: String,
+    end_per_frame_extract: String,
+}
+
+impl ExtractScopeNames {
```

Comment:
Are we having to do the string formats every time we create the job? lazy_static might let us just do it once and reuse.

Reply:
How does `lazy_static` work if the string formats depend on the feature name?

Reply:
yeah, I wasn't thinking through it fully. Maybe it would work if we have a lazy_static created per-feature with the declare_feature macro. I usually don't like doing things in macros, but I think if the alternative is string formatting per frame, I'd definitely consider it. The code might be a little like the RenderFeatureT::render_feature_index() function that returns a global atomic value. If it doesn't work, we can figure something else out.

Reply:
(Feel free to address later, or we can add an issue to try to improve later if you'd rather just leave it for now.)

Reply:
The best idea I've come up with so far is to replace the generated `feature_debug_name` in the macro with `feature_debug_constants` and point to a `const` `static` struct with the `static` strings for the feature name, scope names, etc.

And then if desired, `feature_debug_name()` can be reimplemented just as syntactic sugar around `feature_debug_constants().feature_name`.

Reply:
That sounds good. We might could also have the macro produce functions returning 'static &str for each of these. I do like grouping these under something like feature_debug_constants() to clearly separate what shouldn't be relied upon for correctness.